packages/beacon-node/src/chain/regen/regen.ts

```diff
@@ -195,11 +214,23 @@ export class StateRegenerator implements IStateRegeneratorInternal {
             verifyProposer: false,
             verifySignatures: false,
           },
-          null
+          this.modules.metrics
         );
 
-        // TODO: Persist states, note that regen could be triggered by old states.
-        // Should those take a place in the cache?
+        const stateRoot = toHexString(state.hashTreeRoot());
+        if (b.stateRoot !== stateRoot) {
+          throw new RegenError({
+            slot: b.slot,
+            code: RegenErrorCode.INVALID_STATE_ROOT,
+            actual: stateRoot,
+            expected: b.stateRoot,
+          });
+        }
+
+        if (shouldReload) {
```

Comment:
If `shouldReload` is true, and `getOrReloadLatest()` just retrieves the state from cache without reloading from disk, should we still add state to `stateCache` here? 

In the case of `FIFOBlockStateCache`, if we are adding a state that's already in the cache, it will cause the state to `moveToHead()` or `moveToSecond()`. Conversely, if `shouldReload` is false, why wouldn't we `moveToHead()` or `moveToSecond()`?

Reply:
> If shouldReload is true, and getOrReloadLatest() just retrieves the state from cache without reloading from disk, should we still add state to stateCache here?

in that case this code is not reached because no state is regenerated

> Conversely, if shouldReload is false, why wouldn't we moveToHead() or moveToSecond()?

`shouldReload = false` means a read only operation, it may come from different consumers (for example api or verify attestations) and we really don't want to change the cache due to that. If `shouldReload = true`, that comes from important flow like get state for block processing, or `setHeadState()`, in that case we want to change the cache